infra/batch/datasets/main.py

```diff
@@ -0,0 +1,225 @@
+        dataset_results = connection.execute(select_dataset_statement).all()
+        dataset_id = dataset_results[0][0] if len(dataset_results) > 0 else None
+        dataset_hash = dataset_results[0][1] if len(dataset_results) > 0 else None
+        print(f"Dataset ID = {dataset_id}, Dataset Hash = {dataset_hash}")
+        if dataset_id is None:
+            errors += f"[INTERNAL ERROR] Couldn't find latest dataset related to feed_id {feed_id}\n"
+            return
+
+        # Set the previous version latest field to false
+        if dataset_hash is not None and dataset_hash != md5_file_hash:
+            sql_statement = f"update gtfsdataset set latest=false where id='{dataset_id}'"
+            connection.execute(text(sql_statement))
+
+        sql_statement = f"insert into gtfsdataset (id, feed_id, latest, bounding_box, note, hash, " \
+                        f"download_date, stable_id, hosted_url) " \
+                        f"select '{str(uuid.uuid4())}', feed_id, true, bounding_box, note, " \
+                        f"'{md5_file_hash}', NOW(), stable_id, '{hosted_url}' from " \
+                        f"gtfsdataset where id='{dataset_id}'"
+
+        # In case the dataset doesn't include a hash or the dataset was deleted from the bucket,
+        # update the existing entity
+        if dataset_hash is None or dataset_hash == md5_file_hash:
+            sql_statement = f"update gtfsdataset set hash='{md5_file_hash}', hosted_url='{hosted_url}' where id='{dataset_id}'"
+        connection.execute(text(sql_statement))
+
+        # Commit transaction after every step has run successfully
+        transaction.commit()
+    except Exception as e:
+        if transaction is not None:
+            transaction.rollback()
+        error_traceback = traceback.format_exc()
+        errors += f"[ERROR]: {e}\n{error_traceback}\n"
+        print(f"Logging errors for stable id {stable_id}\n{errors}")
+        storage_client = storage.Client()
+        bucket = storage_client.get_bucket(bucket_name)
+        error_type = 'other'
+        if 'sqlalchemy' in errors:
+            error_type = 'sqlachemy'
+        elif isinstance(e, HTTPError):
+            error_type = f"http/{e.response.status_code}"
+        blob = bucket.blob(f"errors/{datetime.now().strftime('%Y%m%d')}/{error_type}/{stable_id}.log")
+        blob.upload_from_string(errors)
+    finally:
+        if connection is not None:
+            connection.close()
+        gc.collect()  # Free memory
+
+
+def create_bucket(bucket_name):
+    """
+    Creates GCP storage bucket if it doesn't exist
+    :param bucket_name: name of the bucket to create
+    """
+    storage_client = storage.Client()
+    bucket = storage_client.lookup_bucket(bucket_name)
+    if bucket is None:
+        bucket = storage_client.create_bucket(bucket_name)
+        print(f'Bucket {bucket} created.')
+    else:
+        print(f'Bucket {bucket_name} already exists.')
+
+
+def get_db_engine():
```

Comment:
These two functions can be extracted from this files. This will allow other Cloud functions to reuse the same logic